src/motor-operations.cc

```diff
@@ -85,11 +86,40 @@ static char test_acceleration_ok(float acceleration) {
 }
 #endif
 
+typedef struct HistoryPositionInfo {
+  HistoryPositionInfo () : position_steps(0), sign(1) {}
+  int position_steps;
+  uint32_t fraction;
+  signed char sign;
+} HistoryPositionInfo;
+
+// Store the required informations needed to backtrack the absolute position.
+struct MotionQueueMotorOperations::HistorySegment {
+  HistoryPositionInfo pos_info[MOTION_MOTOR_COUNT];
+  unsigned short aux_bits;
+};
+
+MotionQueueMotorOperations::
+  MotionQueueMotorOperations(MotionQueue *backend)
+  : backend_(backend),
+    shadow_queue_(new std::deque<struct HistorySegment>()) {
+  // Initialize the history queue.
+  struct HistorySegment new_hs = {};
+  shadow_queue_->push_front(new_hs);
```

Comment:
> If the machine is not homed yet in the beginning so needs to go backwards - then after homing the absolute positions in our history queue are negative and such, but this is working fine ?

I preferred to keep the position returned absolute also in time so that motor operations do not actually know anything about homing. They just return what they know, the position of the motors since beagleg started.

What I did was retranslating the position using some homing_offset[] array.